File: jw-swarm-nodes/macos/Sources/JWSwarmNode/Tunnel/ConnectionDiagnostics.swift
import Foundation

struct MTLSDiagnosticsResult {
    let success: Bool
    let summary: String
    let details: String
}

enum ConnectionDiagnostics {
    static func run(config: AppConfig) -> MTLSDiagnosticsResult {
        var lines: [String] = []

        let fleet = config.fleet_url.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: fleet), let host = url.host else {
            return MTLSDiagnosticsResult(
                success: false,
                summary: "Invalid Fleet URL",
                details: "Fleet URL is invalid: \(fleet)"
            )
        }

        let port = url.port ?? ((url.scheme == "wss" || url.scheme == "https") ? 443 : 80)
        let tunnelPath = url.path.isEmpty ? "/node/connect" : url.path
        lines.append("Fleet endpoint: \(host):\(port)\(tunnelPath)")

        let nodeCert = config.node_cert
        let caCert = config.ca_cert

        let fm = FileManager.default
        guard fm.fileExists(atPath: nodeCert) else {
            return MTLSDiagnosticsResult(
                success: false,
                summary: "Node certificate missing",
                details: "Configured node cert file does not exist:\n\(nodeCert)"
            )
        }
        guard fm.fileExists(atPath: caCert) else {
            return MTLSDiagnosticsResult(
                success: false,
                summary: "CA certificate missing",
                details: "Configured CA cert file does not exist:\n\(caCert)"
            )
        }

        let nodePEM = (try? String(contentsOfFile: nodeCert, encoding: .utf8)) ?? ""
        let hasClientCert = nodePEM.contains("-----BEGIN CERTIFICATE-----")
        let hasPrivateKey = nodePEM.contains("-----BEGIN PRIVATE KEY-----") ||
            nodePEM.contains("-----BEGIN EC PRIVATE KEY-----") ||
            nodePEM.contains("-----BEGIN RSA PRIVATE KEY-----")

        lines.append("Node PEM: cert block = \(hasClientCert ? "yes" : "no"), private key = \(hasPrivateKey ? "yes" : "no")")
        if !hasClientCert || !hasPrivateKey {
            return MTLSDiagnosticsResult(
                success: false,
                summary: "Node PEM incomplete",
                details: lines.joined(separator: "\n") + "\nExpected both certificate and private key in node PEM."
            )
        }

        let nodeInfo = runOpenSSL(
            args: ["x509", "-in", nodeCert, "-noout", "-subject", "-issuer", "-enddate"],
            timeoutSeconds: 5
        )
        lines.append("\nNode cert:\n\(trim(nodeInfo.output))")

        let caInfo = runOpenSSL(
            args: ["x509", "-in", caCert, "-noout", "-subject", "-issuer", "-enddate"],
            timeoutSeconds: 5
        )
        lines.append("\nCA cert:\n\(trim(caInfo.output))")

        let mtls = runOpenSSL(
            args: [
                "s_client",
                "-connect", "\(host):\(port)",
                "-servername", host,
                "-CAfile", caCert,
                "-cert", nodeCert,
                "-key", nodeCert,
                "-verify_return_error",
            ],
            timeoutSeconds: 8
        )

        let mtlsOutput = trim(mtls.output)
        lines.append("\nmTLS probe:\n\(mtlsOutput)")

        let verified = mtlsOutput.localizedCaseInsensitiveContains("verification: ok") ||
            mtlsOutput.localizedCaseInsensitiveContains("verify return code: 0")
        let connected = mtlsOutput.localizedCaseInsensitiveContains("connection established") ||
            mtlsOutput.localizedCaseInsensitiveContains("connected") ||
            mtlsOutput.contains("CONNECTED(")

        let mtlsOk = mtls.timedOut == false && mtls.exitCode == 0 && verified && connected

        // Verify end-to-end node tunnel path with a real WebSocket upgrade request.
        let tunnelURL: String = {
            var comps = URLComponents()
            comps.scheme = (url.scheme == "wss" || url.scheme == "https") ? "https" : "http"
            comps.host = host
            comps.port = url.port
            comps.path = tunnelPath
            return comps.url?.absoluteString ?? ""
        }()

        let wsProbe = runOpenSSL(
            args: [
                "s_client",
                "-connect", "\(host):\(port)",
                "-servername", host,
                "-CAfile", caCert,
                "-cert", nodeCert,
                "-key", nodeCert,
                "-verify_return_error",
                "-quiet",
            ],
            inputText: [
                "GET \(tunnelPath) HTTP/1.1",
                "Host: \(host)",
                "Connection: Upgrade",
                "Upgrade: websocket",
                // RFC 6455 example nonce (16 bytes base64) to satisfy strict validators.
                "Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==",
                "Sec-WebSocket-Version: 13",
                "",
                "",
            ].joined(separator: "\r\n"),
            timeoutSeconds: 8
        )

        let wsOutput = trim(wsProbe.output)
        lines.append("\nWebSocket upgrade probe (\(tunnelURL)):\n\(wsOutput)")

        let wsAccepted = wsOutput.localizedCaseInsensitiveContains("101 Switching Protocols")
        let wsForbidden = wsOutput.localizedCaseInsensitiveContains(" 403")
        let wsBadRequest = wsOutput.localizedCaseInsensitiveContains(" 400")

        let ok = mtlsOk && wsAccepted
        let summary: String
        if ok {
            summary = "Connection OK (mTLS + tunnel upgrade)"
        } else if mtls.timedOut {
            summary = "Connection test timed out"
        } else if mtls.exitCode != 0 {
            summary = "Connection failed (mTLS handshake failed)"
        } else if wsProbe.timedOut {
            summary = "Connection failed (tunnel upgrade timed out)"
        } else if wsForbidden {
            summary = "Connection failed (tunnel path denied: 403)"
        } else if wsBadRequest {
            summary = "Connection failed (tunnel upgrade rejected: 400)"
        } else {
            summary = "Connection failed (tunnel upgrade not accepted)"
        }

        return MTLSDiagnosticsResult(
            success: ok,
            summary: summary,
            details: lines.joined(separator: "\n")
        )
    }

    private static func trim(_ value: String) -> String {
        value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func runOpenSSL(
        args: [String],
        inputText: String? = nil,
        timeoutSeconds: TimeInterval
    ) -> (exitCode: Int32, output: String, timedOut: Bool) {
        let tool = "/usr/bin/openssl"
        guard FileManager.default.isExecutableFile(atPath: tool) else {
            return (-1, "openssl not available at /usr/bin/openssl", false)
        }

        let process = Process()
        process.executableURL = URL(fileURLWithPath: tool)
        process.arguments = args

        let outPipe = Pipe()
        process.standardOutput = outPipe
        process.standardError = outPipe

        let inputPipe = Pipe()
        process.standardInput = inputPipe

        do {
            try process.run()
            if let inputText {
                if let data = inputText.data(using: .utf8) {
                    inputPipe.fileHandleForWriting.write(data)
                }
            }
            inputPipe.fileHandleForWriting.closeFile()
        } catch {
            return (-1, "Failed to start openssl: \(error.localizedDescription)", false)
        }

        let deadline = Date().addingTimeInterval(timeoutSeconds)
        while process.isRunning && Date() < deadline {
            Thread.sleep(forTimeInterval: 0.05)
        }

        var timedOut = false
        if process.isRunning {
            timedOut = true
            process.terminate()
        }

        let data = (try? outPipe.fileHandleForReading.readToEnd()) ?? Data()
        let output = String(data: data, encoding: .utf8) ?? ""
        return (process.terminationStatus, output, timedOut)
    }
}

File: jw-swarm-nodes/windows/src/JwSwarmNode.Core/Net/ConnectionDiagnostics.cs
using System.Net.Security;
using System.Net.Sockets;
using System.Security.Cryptography.X509Certificates;
using System.Text;
using JwSwarmNode.Core.Config;

namespace JwSwarmNode.Core.Net;

/// <summary>Outcome of a connection diagnostics run.</summary>
public readonly record struct DiagnosticsResult(bool Success, string Summary, string Details);

/// <summary>
/// Connection diagnostics mirroring the macOS <c>ConnectionDiagnostics</c>:
/// verifies cert files exist, the node PEM contains both certificate and
/// private key, the mTLS handshake succeeds against the pinned CA, and the
/// tunnel path accepts a real WebSocket upgrade (HTTP 101).
/// Implemented with pure .NET (<see cref="SslStream"/>) instead of openssl.
/// </summary>
public static class ConnectionDiagnostics
{
    public static DiagnosticsResult Run(NodeConfig config)
    {
        var lines = new List<string>();

        var fleet = config.FleetUrl.Trim();
        if (!Uri.TryCreate(fleet, UriKind.Absolute, out var url) || string.IsNullOrEmpty(url.Host))
        {
            return new DiagnosticsResult(false, "Invalid Fleet URL", $"Fleet URL is invalid: {fleet}");
        }

        var host = url.Host;
        var port = url.IsDefaultPort
            ? (url.Scheme is "wss" or "https" ? 443 : 80)
            : url.Port;
        var tunnelPath = string.IsNullOrEmpty(url.AbsolutePath) || url.AbsolutePath == "/"
            ? "/node/connect"
            : url.AbsolutePath;
        lines.Add($"Fleet endpoint: {host}:{port}{tunnelPath}");

        if (!File.Exists(config.NodeCert))
        {
            return new DiagnosticsResult(false, "Node certificate missing",
                $"Configured node cert file does not exist:\n{config.NodeCert}");
        }
        if (!File.Exists(config.CaCert))
        {
            return new DiagnosticsResult(false, "CA certificate missing",
                $"Configured CA cert file does not exist:\n{config.CaCert}");
        }

        string nodePem;
        try { nodePem = File.ReadAllText(config.NodeCert); }
        catch (Exception ex)
        {
            return new DiagnosticsResult(false, "Node certificate unreadable",
                $"Failed to read node PEM: {ex.Message}");
        }

        var hasClientCert = nodePem.Contains("-----BEGIN CERTIFICATE-----");
        var hasPrivateKey = nodePem.Contains("-----BEGIN PRIVATE KEY-----")
                            || nodePem.Contains("-----BEGIN EC PRIVATE KEY-----")
                            || nodePem.Contains("-----BEGIN RSA PRIVATE KEY-----");
        lines.Add($"Node PEM: cert block = {(hasClientCert ? "yes" : "no")}, private key = {(hasPrivateKey ? "yes" : "no")}");
        if (!hasClientCert || !hasPrivateKey)
        {
            return new DiagnosticsResult(false, "Node PEM incomplete",
                string.Join("\n", lines) + "\nExpected both certificate and private key in node PEM.");
        }

        X509Certificate2 clientCert;
        X509Certificate2 caCert;
        try
        {
            clientCert = X509Certificate2.CreateFromPemFile(config.NodeCert, config.NodeCert);
            lines.Add($"\nNode cert:\nsubject={clientCert.Subject}\nissuer={clientCert.Issuer}\nnotAfter={clientCert.NotAfter:u}");
        }
        catch (Exception ex)
        {
            return new DiagnosticsResult(false, "Node certificate unparsable",
                string.Join("\n", lines) + $"\nFailed to parse node PEM: {ex.Message}");
        }
        try
        {
            caCert = X509Certificate2.CreateFromPem(File.ReadAllText(config.CaCert));
            lines.Add($"\nCA cert:\nsubject={caCert.Subject}\nissuer={caCert.Issuer}\nnotAfter={caCert.NotAfter:u}");
        }
        catch (Exception ex)
        {
            return new DiagnosticsResult(false, "CA certificate unparsable",
                string.Join("\n", lines) + $"\nFailed to parse CA PEM: {ex.Message}");
        }

        // mTLS handshake + WebSocket upgrade probe over a single connection.
        try
        {
            using var tcp = new TcpClient();
            if (!tcp.ConnectAsync(host, port).Wait(TimeSpan.FromSeconds(8)))
            {
                return new DiagnosticsResult(false, "Connection test timed out",
                    string.Join("\n", lines) + $"\nTCP connect to {host}:{port} timed out.");
            }

            using var ssl = new SslStream(tcp.GetStream(), false,
                (_, cert, _, errors) => ValidatePinned(cert, errors, caCert));
            // SslStream requires an exportable key context on Windows.
            using var clientCertForTls = ReExport(clientCert);
            ssl.AuthenticateAsClient(new SslClientAuthenticationOptions
            {
                TargetHost = host,
                ClientCertificates = new X509CertificateCollection { clientCertForTls },
            });
            lines.Add($"\nmTLS probe:\nhandshake OK ({ssl.SslProtocol}, cipher {ssl.NegotiatedCipherSuite})"
                      + $"\nmutual auth = {(ssl.IsMutuallyAuthenticated ? "yes" : "no")}");

            // RFC 6455 example nonce to satisfy strict validators.
            var request =
                $"GET {tunnelPath} HTTP/1.1\r\n" +
                $"Host: {host}\r\n" +
                "Connection: Upgrade\r\n" +
                "Upgrade: websocket\r\n" +
                "Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\n" +
                "Sec-WebSocket-Version: 13\r\n" +
                "\r\n";
            var requestBytes = Encoding.ASCII.GetBytes(request);
            ssl.Write(requestBytes, 0, requestBytes.Length);
            ssl.Flush();

            tcp.ReceiveTimeout = 8000;
            var buffer = new byte[4096];
            var read = ssl.Read(buffer, 0, buffer.Length);
            var response = Encoding.ASCII.GetString(buffer, 0, Math.Max(read, 0));
            var statusLine = response.Split("\r\n").FirstOrDefault() ?? "";
            lines.Add($"\nWebSocket upgrade probe ({tunnelPath}):\n{statusLine}");

            if (response.Contains("101 Switching Protocols", StringComparison.OrdinalIgnoreCase))
            {
                return new DiagnosticsResult(true, "Connection OK (mTLS + tunnel upgrade)",
                    string.Join("\n", lines));
            }

            var summary = response.Contains(" 403", StringComparison.Ordinal)
                ? "Connection failed (tunnel path denied: 403)"
                : response.Contains(" 400", StringComparison.Ordinal)
                    ? "Connection failed (tunnel upgrade rejected: 400)"
                    : "Connection failed (tunnel upgrade not accepted)";
            return new DiagnosticsResult(false, summary, string.Join("\n", lines));
        }
        catch (Exception ex)
        {
            lines.Add($"\nmTLS probe failed: {Unwrap(ex).Message}");
            return new DiagnosticsResult(false, "Connection failed (mTLS handshake failed)",
                string.Join("\n", lines));
        }
    }

    private static bool ValidatePinned(X509Certificate? cert, SslPolicyErrors errors, X509Certificate2 caCert)
    {
        if (errors == SslPolicyErrors.None) return true;
        if (cert is null) return false;
        using var serverCert = new X509Certificate2(cert);
        using var chain = new X509Chain();
        chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
        chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
        chain.ChainPolicy.CustomTrustStore.Add(caCert);
        return chain.Build(serverCert);
    }

    /// <summary>
    /// Round-trips the cert through PKCS#12 so the private key is usable for a
    /// TLS handshake on Windows (ephemeral PEM keys are otherwise rejected).
    /// </summary>
    private static X509Certificate2 ReExport(X509Certificate2 cert)
    {
        try
        {
            return new X509Certificate2(cert.Export(X509ContentType.Pkcs12));
        }
        catch (Exception)
        {
            return cert;
        }
    }

    private static Exception Unwrap(Exception ex) =>
        ex is AggregateException agg && agg.InnerException is not null ? agg.InnerException : ex;
}
